text/0000-prepare-operation.md

```diff
@@ -0,0 +1,374 @@
+
+### Version `io.buildpacks` namespace
+
+Given that the project descriptor's `io.buildpacks` namespace is directly tied to the platform by the fact that it's the platform that needs to consume and parse it, it makes sense to unify the version of the schema with that of the platform API.
+
+
+### Schema
+
+See [Platform Spec changes](#Platform-Spec).
+
+#### Usages Examples
+
+###### Multi-Platform (pack + kpack)
+
+In the following example we show how various platform namespaces can have their own configuration but properties can be promoted to `io.buildpacks.defaults` if the user's intent is to apply it to all platforms.
+
+```toml
+[io.buildpacks]
+schema-version = "0.2"
+
+###
+# common buildpacks config
+##
+
+[io.buildpacks.defaults]
+exclude = ["some-files/**"]
+
+##
+# pack config
+##
+
+[io.buildpacks.pack]
+schema-version = "0.1"
+builder = "cnbs/sample-builder:bionic"
+image = "my-app"                        # (example only)
+run-image = "my-run-image:latest"       # (example only)
+
+##
+# kpack config
+##
+
+[com.vmware.kpack]
+schema-version = "0.1"
+
+[[com.vmware.kpack.build.env]]
+name="CUSTOM_ENV"
+value="SOME_VALUE"
+```
+
+Notice:
+
+1. Both `pack` and `kpack` are expected to apply `io.buildpacks.defaults.exclude`.
+2. `io.buildpacks.pack` has additional properties that are specific to `pack` only.
+
+## Prepare phase
+
+A `prepare` phase would be a new lifecycle phase that gets executed before `create` or `detect` phase. Similar to the `rebaser`, it is outside of the standard [Build operation][build-operation].
+
+[build-operation]: https://github.com/buildpacks/spec/blob/main/platform.md#build
+
+#### Responsibilities
+
+At minimum, the expected reponsibility of the `preparer` would be to:
+
+ - Apply the requested configuration.
+
+#### Inputs
+
+The `prepare` phase would take all the same inputs as `creator` plus the addition of a path to `project.toml`.
+
+#### Outputs
+
+The `prepare` phase may affect the file system, and mutation or create files that would be consumed be futher phases.
+
+## Cloud Native Buildpacks utilities
+
+The Cloud Native Buildpacks project will provide the following functionality in the form of utilities:
+
+###### `go` prepare function
+
+A function that can help developers apply configuration to the filesystem based on a provided configuration. 
+
+> This would be useful for builder providers or platform implementers that choose to provide their own [`preparer` executable](#preparer-executable).
+
+###### `preparer` executable
+
+The Cloud Native Buildpacks project will have a `preparer` it ships along with the existing `lifecyle` image.
+
+> This would be useful for builder providers or platform implementers to not have to develop standard functionality.
+
+The default implementation COULD take care of applying the following configuration:
+
+- `io.buildpacks.defaults.group` →
```

Comment:
Hasn't there been a configuration parameter to replace buildpacks instead of augmenting those present? Was `[build.buildpacks] ` if I am not mistaken. Does that still exist? Should it be added here?

Reply:
I believe you are referring to project descriptor 0.1. This work is based on the [soon-to-release 0.2 descriptor](https://github.com/buildpacks/spec/pull/282) where it's be [relocated to `io.buildpacks.group`](https://github.com/buildpacks/spec/blob/extensions/project-descriptor/0.2/extensions/project-descriptor.md#iobuildpacksgroup-optional).